source/adios2/toolkit/sst/dp/rdma_dp.c

```diff
@@ -617,6 +776,189 @@ static TimestepList GetStep(Rdma_WS_Stream Stream, long Timestep)
     return (Step);
 }
 
+#ifdef SST_HAVE_CRAY_CXI
+int get_cxi_auth_key_from_env(CP_Services Svcs, void *CP_Stream,
+                              struct _SstParams *Params,
+                              struct cxi_auth_key *key,
+                              char **used_device)
+{
+    int vni, first_vni, second_vni, svc_id;
+
+    // Just some safety against string processing.
+    size_t const no_infinite_loops = 10000;
+
+    // struct cxi_auth_key {
+    //     /* The CXI service assigned to the Domain and Endpoints. A CXI
+    //     service
+    //     * is associated with a set of local resource limits, VNIs, and
+    //     Traffic
+    //     * Classes.
+    //     *
+    //     * The svc_id used by an OFI Domain must match all Endpoints belonging
+    //     * to the Domain.
+    //     */
+    //     uint32_t svc_id;
+
+    //     /* The Virtual Network ID (VNI) assigned to the Endpoint. Two
+    //     Endpoints
+    //     * must use the same VNI in order to communicate.
+    //     *
+    //     * Note that while the CXI service may define one or more VNIs which a
+    //     * process can access, an Endpoint is assigned to only one.
+    //     */
+    //     uint16_t vni;
+    // };
+
+    // typical value SLINGSHOT_DEVICES=cxi0,cxi1,cxi2,cxi3
+    char const *slingshot_devices = getenv("SLINGSHOT_DEVICES");
+    char const *preferred_device = get_preferred_domain(Params);
+
+    size_t device_index = 0;
+    for (size_t no_infinite_loop_counter = 0;;
+         ++device_index, ++no_infinite_loop_counter)
+    {
+        if (no_infinite_loop_counter > no_infinite_loops)
+        {
+            return EXIT_FAILURE;
+        }
+
+        int found_end = 0;
+        size_t find_end_of_current_string = 0;
+        for (size_t no_infinite_loop_inner_counter = 0;;
+             ++find_end_of_current_string, ++no_infinite_loop_inner_counter)
+        {
+            if (no_infinite_loop_inner_counter > no_infinite_loops)
+            {
+                return EXIT_FAILURE;
+            }
+
+            switch (slingshot_devices[find_end_of_current_string])
+            {
+            case '\0':
+                found_end = 1;
+                goto break_first_loop;
+            case ',':
+                goto break_first_loop;
```

Comment:
Don't know how strict you guys are with `goto`s. This one here is just to circumvent that you can't break out of a loop while inside a `switch` statement. IMO cleaner and easier to read than the alternatives.

Reply:
I think you could do it the way you are, or you could set a flag inside the case and use it to control breaking out of the loop outside.  That might be less offensive to any anti-goto adherent.

Reply:
That would be the most obvious alternative, yes. But that is more error-prone and more cumbersome to read than the `goto` in my opinion. I think that using `goto` to emulate a labeled break or labeled continue (this exists e.g. in Java) is considered one of the "legitimate" use cases.

Since you're the maintainer, I'm fine with either way. Should I change it?

Reply:
It's fine...